System.Text.Json code generation fails on recursive IReadOnlyDictionary

### Description

System.Text.Json code generation fails with `error CS1031: Type expected` if the serializable class has an `IReadOnlyDictionary` whose value type is that same serializable class.

### Reproduction Steps

Include these two classes in a C# library or console app that references System.Text.Json 6.0.6:
```csharp
public class Dto
{
    public IReadOnlyDictionary<string, Dto>? Data { get; set; }
}

[JsonSerializable(typeof(Dto))]
public partial class JsonContext : JsonSerializerContext
{
}
```

### Expected behavior

Compilation fails with: `...\System.Text.Json.SourceGeneration\System.Text.Json.SourceGeneration.JsonSourceGenerator\JsonContext.IReadOnlyDictionaryStringDto.g.cs(27,128,27,129): error CS1031: Type expected`

That badly generated line of code looks like this: `ObjectCreator = () => new global::System.Collections.Generic.Dictionary<global::System.String, >(),`

### Actual behavior

It should succeed without error, as it does if `Dictionary` is used instead of `IReadOnlyDictionary`.

### Regression?

_No response_

### Known Workarounds

Use `Dictionary` instead of `IReadOnlyDictionary`.

### Configuration

.NET 6.0.401 on Windows x64.

### Other information

_No response_

--- a/src/libraries/System.Text.Json/tests/Common/CollectionTests/CollectionTests.cs
+++ b/src/libraries/System.Text.Json/tests/Common/CollectionTests/CollectionTests.cs
@@ -1,11 +1,48 @@
 // Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Xunit;
+
 namespace System.Text.Json.Serialization.Tests
 {
     public abstract partial class CollectionTests : SerializerTests
     {
         public CollectionTests(JsonSerializerWrapper stringSerializerWrapper)
             : base(stringSerializerWrapper) { }
+
+        [Fact]
+        public async Task RoundtripClassWithRecursiveCollectionProperties()
+        {
+            // Regression test for https://github.com/dotnet/runtime/issues/76802
+
+            var value = new ClassWithRecursiveCollectionTypes
+            {
+                Nested = new(),
+                List = new() { new() },
+                Dictionary = new Dictionary<string, ClassWithRecursiveCollectionTypes>
+                {
+                    ["key"] = new()
+                },
+            };
+
+            string expectedJson = """
+                {
+                    "Nested" : {"Nested":null,"List":null,"Dictionary":null},
+                    "List" : [{"Nested":null,"List":null,"Dictionary":null}],
+                    "Dictionary" : {"key" : {"Nested":null,"List":null,"Dictionary":null}}
+                }
+                """;
+
+            string json = await Serializer.SerializeWrapper(value);
+            JsonTestHelper.AssertJsonEqual(expectedJson, json);
+
+            value = await Serializer.DeserializeWrapper<ClassWithRecursiveCollectionTypes>(json);
+            Assert.NotNull(value.Nested);
+            Assert.Null(value.Nested.Nested);
+            Assert.NotNull(value.List);
+            Assert.NotNull(value.Dictionary);
+        }
     }
 }

--- a/src/libraries/System.Text.Json/tests/Common/TestClasses/TestClasses.cs
+++ b/src/libraries/System.Text.Json/tests/Common/TestClasses/TestClasses.cs
@@ -2286,4 +2286,11 @@ namespace System.Text.Json.Serialization.Tests
             Document.Dispose();
         }
     }
+
+    public class ClassWithRecursiveCollectionTypes
+    {
+        public ClassWithRecursiveCollectionTypes? Nested { get; set; }
+        public List<ClassWithRecursiveCollectionTypes> List { get; set; }
+        public IReadOnlyDictionary<string, ClassWithRecursiveCollectionTypes>? Dictionary { get; set; }
+    }
 }

--- a/src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Tests/Serialization/CollectionTests.cs
+++ b/src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Tests/Serialization/CollectionTests.cs
@@ -453,6 +453,7 @@ namespace System.Text.Json.SourceGeneration.Tests
         [JsonSerializable(typeof(KeyValuePair<int, int>))]
         [JsonSerializable(typeof(KeyValuePair<string, KeyValuePair<string, int>>))]
         [JsonSerializable(typeof(StackWrapper))]
+        [JsonSerializable(typeof(ClassWithRecursiveCollectionTypes))]
         internal sealed partial class CollectionTestsContext_Metadata : JsonSerializerContext
         {
         }
@@ -855,6 +856,7 @@ namespace System.Text.Json.SourceGeneration.Tests
         [JsonSerializable(typeof(KeyValuePair<int, int>))]
         [JsonSerializable(typeof(KeyValuePair<string, KeyValuePair<string, int>>))]
         [JsonSerializable(typeof(StackWrapper))]
+        [JsonSerializable(typeof(ClassWithRecursiveCollectionTypes))]
         internal sealed partial class CollectionTestsContext_Default : JsonSerializerContext
         {
         }

